File: vsh/display.py
from argparse import Namespace
from pathlib import Path
from typing import Callable, Literal, Optional, Self, Tuple

import pyglet
from amaranth import Elaboratable, Signal
from amaranth.sim import Simulator
from pyglet import gl
from pyglet.image import ImageData
from pyglet.window import Window, key

from oled import OLED
from oled.sh1107 import Base, Cmd, DataBytes
from .display_base import DisplayBase
from .oled_connector import OLEDConnector
from .switch_connector import SwitchConnector

__all__ = ["run"]


def run(top: Elaboratable, args: Namespace):
    simulator = Simulator(top)

    v = Display(top, simulator)

    # 1Mbps on a 4MHz system clock gets us the sweet spot
    # where the I2C clock (2MHz) is at a 1:2 ratio with the
    # system clock.  This means every tick is either "half"
    # or "full".
    simulator.add_clock(1 / 4e6)
    if v.switch_connector:
        simulator.add_sync_process(v.switch_connector.sim_process)
    if v.oled_connector:
        simulator.add_sync_process(v.oled_connector.sim_process)

    if args.vcd:
        vcd_path = Path(__file__).parent.parent / "build" / "vsh.vcd"
        with simulator.write_vcd(str(vcd_path)):
            v.run()
    else:
        v.run()


class Display(DisplayBase, Window):
    top: Elaboratable
    simulator: Simulator
    press_button: bool

    switch_connector: Optional[SwitchConnector]
    oled_connector: Optional[OLEDConnector]

    power: bool
    dcdc: bool  # NE
    dclk_freq: Cmd.SetDisplayClockFrequency.Freq  # NE
    dclk_ratio: int  # NE
    precharge_period: int  # NE
    discharge_period: int  # NE
    vcom_desel: int  # TODO
    all_on: bool  # TODO
    reversed: bool  # TODO
    contrast: int  # TODO
    start_line: int  # TODO
    start_column: int  # TODO
    page_address: int
    column_address: int
    addressing_mode: Cmd.SetMemoryAddressingMode.Mode
    multiplex: int  # XXX
    segment_remap: bool
    com_scan_reversed: bool

    idata: bytearray
    img: ImageData
    img_stale: bool

    def __init__(self, top: Elaboratable, simulator: Simulator):
        super().__init__(
            width=self.WINDOW_WIDTH,
            height=self.WINDOW_HEIGHT,
            caption="SH1107 OLED I²C",
        )

        self.top = top
        self.simulator = simulator

        switch = getattr(top, "switch", None)
        if switch is not None:
            assert isinstance(switch, Signal)
            self.switch_connector = SwitchConnector(switch)

        oled = getattr(top, "oled", None)
        if oled is not None:
            assert isinstance(oled, OLED)
            self.oled_connector = OLEDConnector(oled, self.process_i2c)

        self.power = False
        self.dcdc = True
        self.dclk_freq = Cmd.SetDisplayClockFrequency.Freq.Zero
        self.dclk_ratio = 1
        self.precharge_period = 2
        self.discharge_period = 2
        self.vcom_desel = 0x35
        self.all_on = False
        self.reversed = False
        self.contrast = 0x80
        self.start_line = 0
        self.start_column = 0
        self.page_address = 0
        self.column_address = 0
        self.addressing_mode = Cmd.SetMemoryAddressingMode.Mode.Page
        self.multiplex = 128
        self.segment_remap = False
        self.com_scan_reversed = False

        self.idata = bytearray(self.BLACK * self.I2C_WIDTH * self.I2C_HEIGHT)
        self.img = ImageData(self.I2C_WIDTH, self.I2C_HEIGHT, "RGBA", bytes(self.idata))
        self.img_stale = False

    def run(self):
        pyglet.app.run()

    def on_draw(self):  # pyright: reportIncompatibleMethodOverride=none
        # TODO(ari): range(?)
        for _ in range(1000):
            self.simulator.advance()

        self.voyager2.blit(0, 0, width=self.WINDOW_WIDTH, height=self.WINDOW_HEIGHT)
        self._draw_top()
        self._draw_oled()

    TOP_COLS: list[list[Tuple[str, Callable[[Self], bool | str]]]] = [
        [
            ("power on", lambda d: d.power),
            ("dc/dc on", lambda d: d.dcdc),
            ("dclk", lambda d: f"{int(d.dclk_freq)}% {d.dclk_ratio}x"),
            ("pre/dis", lambda d: f"{d.precharge_period}/{d.discharge_period}"),
            ("vcom desel", lambda d: f"{d.vcom_desel:02x}"),
        ],
        [
            ("all on", lambda d: d.all_on),
            ("reversed", lambda d: d.reversed),
            ("contrast", lambda d: f"{d.contrast:02x}"),
        ],
        [
            ("start", lambda d: f"{d.start_line:02x}/{d.start_column:02x}"),
            ("address", lambda d: f"{d.page_address:02x}/{d.column_address:02x}"),
            ("mode", lambda d: "page" if d.addressing_mode == 0 else "column"),
            ("multiplex", lambda d: f"{d.multiplex:02x}"),
        ],
        [
            ("seg remap", lambda d: d.segment_remap),
            ("com rev", lambda d: d.com_scan_reversed),
        ],
    ]

    def _draw_top(self):
        left = self.PADDING

        for col in self.TOP_COLS:
            top = self.WINDOW_HEIGHT - self.PADDING
            for name, lamb in col:
                x = left

                val: bool | str = lamb(self)
                if isinstance(val, bool):
                    x += self.CHECKBOX_SIZE + self.CHECKBOX_TEXT_GAP
                    self.render_text(
                        name, self.WHITE, x=x, y=top, anchor_y="center", bold=val
                    ).draw()
                    (self.checked if val else self.unchecked).blit(
                        left, top - self.CHECKBOX_DOWN
                    )
                else:
                    label = self.render_text(
                        f"{name}: <b>{val}</b>",
                        self.WHITE,
                        x=x,
                        y=top,
                        anchor_y="center",
                    )
                    label.draw()

                top -= self.TOP_ROW_HEIGHT

            left += self.TOP_COL_WIDTH

    def _draw_oled(self):
        gl.glEnable(gl.GL_BLEND)
        gl.glBlendFunc(gl.GL_SRC_ALPHA, gl.GL_ONE_MINUS_SRC_ALPHA)

        if self.img_stale:
            self.img.set_data("RGBA", 4 * self.I2C_WIDTH, bytes(self.idata))
            self.img_stale = False

        if self.power:
            self.on_border.blit(
                self.PADDING,
                self.PADDING,
                width=self.BORDER_FILL_WIDTH,
                height=self.BORDER_FILL_HEIGHT,
            )
            self.img.blit(
                x=(
                    self.PADDING
                    + self.BORDER_WIDTH
                    + (
                        self.I2C_WIDTH * self.DISPLAY_SCALE
                        if self.com_scan_reversed
                        else 0
                    )
                ),
                y=(
                    self.PADDING
                    + self.BORDER_WIDTH
                    + self.I2C_HEIGHT * self.DISPLAY_SCALE
                ),
                width=(
                    self.I2C_WIDTH
                    * self.DISPLAY_SCALE
                    * (-1 if self.com_scan_reversed else 1)
                ),
                height=(self.I2C_HEIGHT * self.DISPLAY_SCALE * (-1)),
            )
        else:
            self.off_border.blit(
                self.PADDING,
                self.PADDING,
                width=self.BORDER_FILL_WIDTH,
                height=self.BORDER_FILL_HEIGHT,
            )

    def set_px(self, x: int, y: int, val: Literal[0, 1]):
        off = 4 * (y * self.I2C_HEIGHT + x)
        self.idata[off : off + 4] = self.WHITE if val else self.BLACK
        self.img_stale = True

    def process_i2c(self, msg: list[Base | DataBytes]):
        for c in msg:
            match c:
                case Cmd.SetLowerColumnAddress(lower=lower):
                    self.column_address = (self.column_address & 0xF0) | lower

                case Cmd.SetHigherColumnAddress(higher=higher):
                    self.column_address = (self.column_address & 0x0F) | (higher << 4)

                case Cmd.SetMemoryAddressingMode(mode=mode):
                    self.addressing_mode = mode

                case Cmd.SetContrastControlRegister(level=level):
                    self.contrast = level

                case Cmd.SetSegmentRemap(adc=adc):
                    self.segment_remap = adc == Cmd.SetSegmentRemap.Adc.Flipped

                case Cmd.SetMultiplexRatio(ratio=ratio):
                    self.multiplex = ratio

                case Cmd.SetEntireDisplayOn(on=on):
                    self.all_on = on

                case Cmd.SetDisplayReverse(reverse=reverse):
                    self.reversed = reverse

                case Cmd.SetDisplayOffset(offset=offset):
                    self.start_line = offset

                case Cmd.SetDCDC(on=on):
                    self.dcdc = on

                case Cmd.DisplayOn(on=on):
                    self.power = on

                case Cmd.SetPageAddress(page=page):
                    self.page_address = page

                case Cmd.SetCommonOutputScanDirection(direction=direction):
                    self.com_scan_reversed = (
                        direction
                        == Cmd.SetCommonOutputScanDirection.Direction.Backwards
                    )

                case Cmd.SetDisplayClockFrequency(ratio=ratio, freq=freq):
                    self.dclk_freq = freq
                    self.dclk_ratio = ratio

                case Cmd.SetPreDischargePeriod(
                    precharge=precharge, discharge=discharge
                ):
                    self.precharge_period = precharge
                    self.discharge_period = discharge

                case Cmd.SetVCOMDeselectLevel(level=level):
                    self.vcom_desel = level

                case Cmd.SetDisplayStartColumn(column=column):
                    self.start_column = column

                case Cmd.ReadModifyWrite():
                    raise NotImplementedError

                case Cmd.End():
                    raise NotImplementedError

                case Cmd.Nop():
                    pass

                case DataBytes(data=data):
                    page_count = self.I2C_HEIGHT // 8
                    for b in data:
                        for i in range(7, -1, -1):
                            if not self.segment_remap:
                                pa = self.page_address * 8 + i
                            else:
                                pa = (page_count - self.page_address - 1) * 8 + (7 - i)
                            self.set_px(
                                self.column_address,
                                pa,
                                1 if ((b >> i) & 0x01) == 0x01 else 0,
                            )
                        if (
                            self.addressing_mode
                            == Cmd.SetMemoryAddressingMode.Mode.Page
                        ):
                            self.column_address = (
                                self.column_address + 1
                            ) % self.I2C_WIDTH
                        else:
                            self.page_address = (self.page_address + 1) % page_count

                case Base():
                    assert False

    def on_key_press(self, symbol: int, modifiers: int):
        if symbol == key.ESCAPE:
            self.dispatch_event("on_close")
            return

        if symbol == key.RETURN:
            assert self.switch_connector
            self.switch_connector.press()

File: vsh/src/Display.zig
const std = @import("std");
const gk = @import("gamekit");

const DisplayBase = @import("./DisplayBase.zig");
const FPGAThread = @import("./FPGAThread.zig");
const SH1107 = @import("./SH1107.zig");

const Display = @This();

fpga_thread: *FPGAThread,
base: DisplayBase,
img: gk.gfx.Texture,

pub fn init() !Display {
    const fpga_thread = try FPGAThread.start();
    const base = try DisplayBase.init();
    const img = gk.gfx.Texture.init(DisplayBase.i2c_width, DisplayBase.i2c_height);

    return .{
        .fpga_thread = fpga_thread,
        .base = base,
        .img = img,
    };
}

pub fn deinit(self: Display) void {
    self.fpga_thread.stop();
}

pub fn update(self: *Display) bool {
    if (gk.input.keyPressed(.escape)) {
        return false;
    }

    inline for (.{ .num_1, .num_2, .num_3, .num_4, .num_5, .num_6, .num_7, .num_8, .num_9, .num_0 }, 1..) |n, i| {
        if (gk.input.keyPressed(n)) {
            self.fpga_thread.press_switch_connector(i);
        }
    }

    return true;
}

pub fn render(self: *Display) void {
    const gfx = gk.gfx;

    gfx.beginPass(.{ .color = gk.math.Color.black });

    gfx.draw.tex(self.base.voyager2, .{ .x = 0, .y = 0 });

    const sh1107 = self.fpga_thread.acquire_sh1107();
    self.drawTop(&sh1107);
    self.drawOLED(&sh1107);

    gfx.endPass();
}

const TopDrawState = struct {
    display: *Display,

    left: f32,
    top: f32,

    pub fn start(display: *Display) TopDrawState {
        return .{
            .display = display,
            .left = @floatFromInt(f32, DisplayBase.padding),
            .top = @floatFromInt(f32, DisplayBase.padding),
        };
    }

    pub fn check(self: *TopDrawState, name: []const u8, value: bool) void {
        gk.gfx.draw.textOptions(name, self.display.base.fontbook, .{
            .x = self.left + @floatFromInt(f32, DisplayBase.checkbox_size + DisplayBase.checkbox_text_gap),
            .y = self.top,
            .alignment = .left_middle,
            .color = DisplayBase.white,
        });
        if (value) {
            // "bold"
            gk.gfx.draw.textOptions(name, self.display.base.fontbook, .{
                .x = self.left + @floatFromInt(f32, DisplayBase.checkbox_size + DisplayBase.checkbox_text_gap) + 0.5,
                .y = self.top,
                .alignment = .left_middle,
                .color = DisplayBase.white,
            });
        }

        const y = self.top - @floatFromInt(f32, DisplayBase.checkbox_up);

        if (value) {
            gk.gfx.draw.rect(
                .{
                    .x = self.left,
                    .y = y,
                },
                DisplayBase.checkbox_size,
                DisplayBase.checkbox_size,
                DisplayBase.white,
            );
            gk.gfx.draw.hollowRect(
                .{
                    .x = self.left,
                    .y = y,
                },
                DisplayBase.checkbox_size,
                DisplayBase.checkbox_size,
                1,
                DisplayBase.on_border,
            );
        } else {
            gk.gfx.draw.rect(
                .{
                    .x = self.left,
                    .y = y,
                },
                DisplayBase.checkbox_size,
                DisplayBase.checkbox_size,
                DisplayBase.off,
            );
            gk.gfx.draw.hollowRect(
                .{
                    .x = self.left,
                    .y = y,
                },
                DisplayBase.checkbox_size,
                DisplayBase.checkbox_size,
                1,
                DisplayBase.off_border,
            );
        }

        self.top += @floatFromInt(f32, DisplayBase.top_row_height);
    }

    pub fn fmt(self: *TopDrawState, name: []const u8, comptime f: []const u8, args: anytype) void {
        var buf1: [64]u8 = undefined;
        var buf2: [64]u8 = undefined;
        const s1 = std.fmt.bufPrint(&buf1, f, args) catch unreachable;
        const s2 = std.fmt.bufPrint(&buf2, "{s}: {s}", .{ name, s1 }) catch unreachable;
        gk.gfx.draw.textOptions(s2, self.display.base.fontbook, .{
            .x = self.left,
            .y = self.top,
            .alignment = .left_middle,
            .color = DisplayBase.white,
        });
        self.top += @floatFromInt(f32, DisplayBase.top_row_height);
    }

    pub fn row(self: *TopDrawState) void {
        self.left += @floatFromInt(f32, DisplayBase.top_col_width);
        self.top = @floatFromInt(f32, DisplayBase.padding);
    }
};

fn drawTop(self: *Display, sh1107: *const SH1107) void {
    var tds = TopDrawState.start(self);
    tds.check("power on", sh1107.power);
    tds.check("dc/dc on", sh1107.dcdc);
    tds.fmt("dclk", "{d}% {d}x", .{ @intFromEnum(sh1107.dclk_freq), sh1107.dclk_ratio });
    tds.fmt("pre/dis", "{d}/{d}", .{ sh1107.precharge_period, sh1107.discharge_period });
    tds.fmt("vcom desel", "{x:0>2}", .{sh1107.vcom_desel});

    tds.row();
    tds.check("all on", sh1107.all_on);
    tds.check("reversed", sh1107.reversed);
    tds.fmt("contrast", "{x:0>2}", .{sh1107.contrast});

    tds.row();
    tds.fmt("start", "{x:0>2}/{x:0>2}", .{ sh1107.start_line, sh1107.start_line });
    tds.fmt("address", "{x:0>2}/{x:0>2}", .{ sh1107.page_address, sh1107.column_address });
    tds.fmt("mode", "{s}", .{sh1107.addressing_mode.str()});
    tds.fmt("multiplex", "{x:0>2}", .{sh1107.multiplex});

    tds.row();
    tds.fmt("offset", "{x:0>2}", .{sh1107.start_offset});
    tds.fmt("start", "{x:0>2}", .{sh1107.start_line});
    tds.check("seg remap", sh1107.segment_remap == .Flipped);
    tds.check("com rev", sh1107.com_scan_dir == .Backwards);
}

fn dtStart(self: *Display) void {
    self.base.dtStart();
}

fn drawOLED(self: *Display, sh1107: *const SH1107) void {
    if (self.fpga_thread.idata_stale.compareAndSwap(true, false, .Acquire, .Monotonic) == null) {
        self.fpga_thread.idata_mutex.lock();
        defer self.fpga_thread.idata_mutex.unlock();
        self.img.setData(gk.math.Color, &self.fpga_thread.idata);
    }

    if (sh1107.power) {
        gk.gfx.draw.hollowRect(
            .{
                .x = @floatFromInt(f32, DisplayBase.padding),
                .y = @floatFromInt(f32, DisplayBase.padding + DisplayBase.top_area),
            },
            DisplayBase.border_fill_width,
            DisplayBase.border_fill_height,
            DisplayBase.border_width,
            DisplayBase.on_border,
        );

        const x_factor = if (sh1107.com_scan_dir == .Backwards) @as(f32, -1) else @as(f32, 1);
        const x_scale = @floatFromInt(f32, DisplayBase.display_scale) * x_factor;
        const shift = if (sh1107.com_scan_dir == .Backwards) @floatFromInt(f32, DisplayBase.i2c_width) else 0;
        const display_scale = @floatFromInt(f32, DisplayBase.display_scale);

        const start_line = sh1107.start_line +% sh1107.start_offset;

        gk.gfx.draw.texScaleXYRegionAngle(
            self.img,
            .{
                .x = @floatFromInt(f32, DisplayBase.padding + DisplayBase.border_width) + @floatFromInt(f32, DisplayBase.i2c_width) * display_scale,
                .y = @floatFromInt(f32, DisplayBase.padding + DisplayBase.border_width + DisplayBase.top_area) + shift * display_scale,
            },
            .{
                .x = @floatFromInt(f32, start_line),
                .y = 0,
                .w = @floatFromInt(f32, DisplayBase.i2c_width - start_line),
                .h = @floatFromInt(f32, DisplayBase.i2c_height),
            },
            x_scale,
            DisplayBase.display_scale,
            std.math.pi * 0.5,
        );
        if (start_line > 0) {
            gk.gfx.draw.texScaleXYRegionAngle(
                self.img,
                .{
                    .x = @floatFromInt(f32, DisplayBase.padding + DisplayBase.border_width) + @floatFromInt(f32, DisplayBase.i2c_width) * display_scale,
                    .y = @floatFromInt(f32, DisplayBase.padding + DisplayBase.border_width + DisplayBase.top_area) + (shift + @floatFromInt(f32, DisplayBase.i2c_width - start_line) * x_factor) * display_scale,
                },
                .{
                    .x = 0,
                    .y = 0,
                    .w = @floatFromInt(f32, start_line),
                    .h = @floatFromInt(f32, DisplayBase.i2c_height),
                },
                x_scale,
                DisplayBase.display_scale,
                std.math.pi * 0.5,
            );
        }
    } else {
        gk.gfx.draw.rect(
            .{
                .x = @floatFromInt(f32, DisplayBase.padding),
                .y = @floatFromInt(f32, DisplayBase.padding + DisplayBase.top_area),
            },
            DisplayBase.border_fill_width,
            DisplayBase.border_fill_height,
            DisplayBase.off,
        );
        gk.gfx.draw.hollowRect(
            .{
                .x = @floatFromInt(f32, DisplayBase.padding),
                .y = @floatFromInt(f32, DisplayBase.padding + DisplayBase.top_area),
            },
            DisplayBase.border_fill_width,
            DisplayBase.border_fill_height,
            DisplayBase.border_width,
            DisplayBase.off_border,
        );
    }
}
